tsweb/tsweb.go

```diff
@@ -137,6 +139,61 @@ func stripPort(hostport string) string {
 	return net.JoinHostPort(host, "443")
 }
 
+// Handler is like net/http.Handler, but the handler can return an
+// error.
+type Handler interface {
+	ServeHTTP(http.ResponseWriter, *http.Request) error
+}
+
+// handler is an http.Handler that wraps a Handler and handles errors.
+type handler struct {
+	h    Handler
+	logf logger.Logf
+}
+
+func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	msg := Msg{
+		Where: "http",
+		When:  time.Now(),
+		HTTP: &MsgHTTP{
+			Path:       r.URL.Path,
+			RemoteAddr: r.RemoteAddr,
+			UserAgent:  r.UserAgent(),
+			Referer:    r.Referer(),
+		},
+	}
+	err := h.h.ServeHTTP(w, r)
+	if err == context.Canceled {
+		// No need to inform client, but still log the
+		// cancellation.
+		msg.HTTP.Code = 499 // nginx convention: Client Closed Request
+		msg.Err = err
+	} else if hErr, ok := err.(HTTPError); ok {
+		msg.HTTP.Code = hErr.Code
+		msg.Err = hErr.Err
+		msg.Msg = hErr.Msg
+		http.Error(w, hErr.Msg, hErr.Code)
+	} else if err != nil {
+		msg.HTTP.Code = 500
+		msg.Err = err
+		msg.Msg = "internal server error"
+		http.Error(w, msg.Msg, msg.HTTP.Code)
+	}
+
+	// TODO(danderson): needed? Copied from existing code, but
+	// doesn't HTTPServer do this by itself?
+	if f, _ := w.(http.Flusher); f != nil {
+		f.Flush()
+	}
+	h.logf("%s", msg)
+}
+
+// ErrHandler returns an http.Handler that logs requests and handles
```

Comment:
I don't know how to read this. How does it handle them?

Reply:
I think I'd just rename this to be, say, `StdHandler`. It converts the ErrHandler into a std Handler. As written it sounds like it returns an Err Handler.